core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
       throw new RuntimeException(e);
     }
 
+    startMasters(false);
     while (!Thread.interrupted()) {
-      startMasters(false);
-      LOG.info("Secondary started");
       mLeaderSelector.waitForState(State.PRIMARY);
+      if (gainPrimacy()) {
+        mLeaderSelector.waitForState(State.SECONDARY);
+        losePrimacy();
+      }
+    }
+  }
+
+  /**
+   * @return true if the master successfully upgrades to primary
+   */
+  private boolean gainPrimacy() throws Exception {
+    AtomicBoolean lostPrimacy = new AtomicBoolean(false);
+    try (Scoped scoped = mLeaderSelector.onStateChange(state -> lostPrimacy.set(true))) {
```

Comment:
does this have to check what the `state` is?

Reply:
master should only complete the upgrade to primary if it stays primary throughout the time it takes to call `journalSystem.gainPrimacy()`. I've renamed this variable to `unstable` to make it clearer